tools/box/box.h

```diff
@@ -53,38 +56,122 @@ jxl::Status ParseBoxHeader(const uint8_t** next_in, size_t* available_in,
-
-  // Brotli-compressed exif metadata, if present. The data points to the brotli
-  // compressed stream, it is not decompressed here.
-  const uint8_t* exfc = nullptr;  // Not owned
-  size_t exfc_size = 0;
-
-  // XML boxes for XMP. There may be multiple XML boxes.
-  // Each entry points to XML location and provides size.
-  // The memory is not owned.
-  // TODO(lode): for C API, cannot use std::vector.
-  std::vector<std::pair<const uint8_t*, size_t>> xml;
-
-  // Brotli-compressed xml boxes. The bytes are given in brotli-compressed form
-  // and are not decompressed here.
-  std::vector<std::pair<const uint8_t*, size_t>> xmlc;
-
-  // JUMBF superbox data, or null if not present in the container.
-  // The parsing of the nested boxes inside is not handled here.
-  const uint8_t* jumb = nullptr;  // Not owned
-  size_t jumb_size = 0;
-
+  std::vector<BrobBlob> blobs;
+
+  // Returns a pointer to a BrobBlob with uncompressed data (or nullptr if not
+  // found). In case there are multiple of that type, return the index-th one.
+  const BrobBlob* getBlob(const char* type, size_t index = 0) const {
+    for (auto& b : blobs) {
+      if (!memcmp(type, b.type, 4)) {
+        if (index > 0) {
+          index--;
+          continue;
+        }
+        if (b.udata_size == 0) {
+          BrotliDecoderState* brotli_dec =
+              BrotliDecoderCreateInstance(nullptr, nullptr, nullptr);
+          struct BrotliDecDeleter {
+            BrotliDecoderState* brotli_dec;
+            ~BrotliDecDeleter() { BrotliDecoderDestroyInstance(brotli_dec); }
+          } brotli_dec_deleter{brotli_dec};
+          size_t available_in = b.cdata_size;
+          b.pbytes.resize(available_in);
+          size_t available_out = b.pbytes.size();
+          const uint8_t* in = b.cdata;
+          uint8_t* out = b.pbytes.data();
+          size_t tot_dec = 0;
+          while (available_in > 0 || !BrotliDecoderIsFinished(brotli_dec)) {
+            BrotliDecoderResult result = BrotliDecoderDecompressStream(
+                brotli_dec, &available_in, &in, &available_out, &out, &tot_dec);
+            if (result ==
+                BrotliDecoderResult::BROTLI_DECODER_RESULT_NEEDS_MORE_OUTPUT) {
+              b.pbytes.resize(b.pbytes.size() + 10240);
+              out = b.pbytes.data() + tot_dec;
+              available_out = b.pbytes.size() - tot_dec;
+            } else if (result !=
+                       BrotliDecoderResult::BROTLI_DECODER_RESULT_SUCCESS) {
+              JXL_WARNING("Brotli decoding error: %s\n",
+                          BrotliDecoderErrorString(
+                              BrotliDecoderGetErrorCode(brotli_dec)));
+              return nullptr;
+            }
+          }
+          b.udata = b.pbytes.data();
+          b.udata_size = tot_dec;
+        }
+        return &b;
+      }
+    }
+    return nullptr;
+  }
+
+  void addBlob(const char* type, const uint8_t* data, size_t data_size,
+               bool compress = true) {
```

Comment:
I'm not sure if compress should be the default, maybe just don't have a default. While compressing is the best idea, there may also be cases where you want it uncompressed to be compatible with exif, jumb, ... tools that don't support this. Since addBlob looks generic for adding any type of box, it might also be used to add small application-specific boxes where compression wouldn't help. WDYT?

Reply:
Makes sense, I'll remove the default. For now, I suppose there are no tools that can interpret compressed exif/xmp so it would probably be useful to have a cjxl option to produce uncompressed boxes, in case compatibility with existing tools is important.